samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManagerUtil.java

```diff
@@ -411,4 +438,292 @@ public static Set<String> getSideInputStoreNames(
+              String msg = String.format("Error restoring state for task: %s", taskName);
+              LOG.error(msg, ex);
+              throw new SamzaException(msg, ex); // wrap in unchecked exception to throw from lambda
+            }
+          }
+
+          // Stop all persistent stores after restoring. Certain persistent stores opened in BulkLoad mode are compacted
+          // on stop, so paralleling stop() also parallelizes their compaction (a time-intensive operation).
+          try {
+            taskRestoreManager.close();
+          } catch (Exception e) {
+            LOG.error("Error closing restore manager for task: {} after {} restore", taskName,
+                ex != null ? "unsuccessful" : "successful", e);
+            // ignore exception from close. container may still be able to continue processing/backups
+            // if restore manager close fails.
+          }
+          return null;
+        });
+        taskRestoreFutures.add(taskRestoreFuture);
+      });
+    });
+
+    return taskRestoreFutures;
+  }
+
+  public static CompletableFuture<Void> restoreTaskStoresAndCreateCheckpointWithGetDeleted(TaskName taskName,
+      Map<TaskName, Checkpoint> taskCheckpoints, CheckpointManager checkpointManager,
+      TaskRestoreManager taskRestoreManager, Config config, Map<TaskName, TaskInstanceMetrics> taskInstanceMetrics,
+      ExecutorService executor, Set<String> storesToRestore, File loggedStoreBaseDirectory, JobContext jobContext,
+      ContainerModel containerModel) {
+
+    // if taskInstanceMetrics are specified use those for store metrics,
+    // otherwise (in case of StorageRecovery) use a blank MetricsRegistryMap
+    MetricsRegistry metricsRegistry =
+        taskInstanceMetrics.get(taskName) != null ? taskInstanceMetrics.get(taskName).registry()
+            : new MetricsRegistryMap();
+
+    BlobStoreManager blobStoreManager = getBlobStoreManager(config, executor);
+    JobConfig jobConfig = new JobConfig(config);
+    BlobStoreUtil blobStoreUtil =
+        new BlobStoreUtil(blobStoreManager, executor, new BlobStoreConfig(config), null,
+            new BlobStoreRestoreManagerMetrics(metricsRegistry));
+
+    BlobStoreRestoreManager blobStoreRestoreManager = (BlobStoreRestoreManager) taskRestoreManager;
+
+    CheckpointId checkpointId = CheckpointId.create();
+    Map<String, String> oldSCMs = ((CheckpointV2) taskCheckpoints.get(taskName))
+        .getStateCheckpointMarkers().get(BlobStoreStateBackendFactory.class.getName());
+
+    // Returns a single future that guarantees all the following are completed, in this order:
+    // 1. Restore state locally by getting deleted blobs from the blob store.
+    // 2. Create a new Checkpoint from restored state and back it up on the blob store.
+    // 3. Clean up old/deleted Snapshot
+    // 4. Remove TTL from the new Snapshot on the blob store
+    // 5. Write the new checkpoint to checkpoint topic and update taskCheckpoints map
```

Comment:
Is this mutating the map passed in all the way from CSM? If so, strongly prefer returning new values instead of mutating params / incoming data structures.

Reply:
Updated that to not mutate anymore.